src/codecs/gif.rs

```diff
@@ -154,7 +168,17 @@ impl<'a, R: 'a + Read> ImageDecoder<'a> for GifDecoder<R> {
         } else {
             // If the frame does not match the logical screen, read into an extra buffer
             // and 'insert' the frame from left/top to logical screen width/height.
-            let mut frame_buffer = vec![0; self.reader.buffer_size()];
+            let buffer_size = self.reader.buffer_size();
+
+            self.limits.reserve_usize(buffer_size)?;
+
+            let mut frame_buffer = vec![0; buffer_size];
+
+            self.limits.free(
```

Comment:
I think the reserve/free API is good, but I needed to `free` earlier than this buffer is actually freed to avoid the limits not actually being reset.

Maybe an alternate API like a `Limits` being an Allocator wrapper would be useful. Not sure! Probably best to just get more examples of code that use the limits API and then see what fits them the best.

Reply:
It's probably correct, an allocator API would be interesting but it would need to interact with all containers. `Vec` and `Box` is potentially a good start but not generic enough. But it's hard to say if it was better to wait on an the allocator interface or to solve it separately.

Given that there is no concurrency in this specific, I would say the order doesn't matter to here. The only effect of free is to allow following `reserve` calls to succeed.